plugin/hijack.go

```diff
@@ -0,0 +1,205 @@
+	ErrCannotHijack    = errors.New("response cannot be hijacked")
+)
+
+func (w *httpResponseWriterRPCServer) HjConnRWRead(b []byte, reply *[]byte) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	data := make([]byte, len(b))
+	n, err := w.hjr.brw.Read(data)
+	if err != nil {
+		return err
+	}
+	*reply = data[:n]
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnRWWrite(b []byte, reply *int) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	n, err := w.hjr.brw.Write(b)
+	if err != nil {
+		return err
+	}
+	*reply = n
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnRead(size int, reply *[]byte) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	if len(w.hjr.readBuf) < size {
+		w.hjr.readBuf = make([]byte, size)
+	}
+	n, err := w.hjr.conn.Read(w.hjr.readBuf[:size])
+	if err != nil {
+		return err
+	}
+	*reply = w.hjr.readBuf[:n]
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnWrite(b []byte, reply *int) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	n, err := w.hjr.conn.Write(b)
+	if err != nil {
+		return err
+	}
+	*reply = n
+	return nil
+}
+
+func (w *httpResponseWriterRPCServer) HjConnClose(args struct{}, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.Close()
+}
+
+func (w *httpResponseWriterRPCServer) HjConnSetDeadline(t time.Time, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.SetDeadline(t)
+}
+
+func (w *httpResponseWriterRPCServer) HjConnSetReadDeadline(t time.Time, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.SetReadDeadline(t)
+}
+
+func (w *httpResponseWriterRPCServer) HjConnSetWriteDeadline(t time.Time, reply *struct{}) error {
+	if w.hjr == nil {
+		return ErrNotHijacked
+	}
+	return w.hjr.conn.SetWriteDeadline(t)
+}
+
+func (w *httpResponseWriterRPCServer) HijackResponse(args struct{}, reply *struct{}) error {
+	if w.hjr != nil {
+		return ErrAlreadyHijacked
+	}
+	hj, ok := w.w.(http.Hijacker)
+	if !ok {
+		return ErrCannotHijack
+	}
+	conn, brw, err := hj.Hijack()
+	if err != nil {
+		return err
+	}
+
+	w.hjr = &hijackedResponse{
+		conn:    conn,
+		brw:     brw,
+		readBuf: make([]byte, hijackedConnReadBufSize),
+	}
+	return nil
+}
+
+type hijackedConn struct {
+	client *rpc.Client
+}
+
+type hijackedConnRW struct {
+	client *rpc.Client
+}
+
+func (w *hijackedConnRW) Read(b []byte) (int, error) {
+	var data []byte
+	if err := w.client.Call("Plugin.HjConnRWRead", b, &data); err != nil {
+		return 0, err
+	}
+	copy(b, data)
+	return len(data), nil
+}
+
+func (w *hijackedConnRW) Write(b []byte) (int, error) {
+	var n int
+	if err := w.client.Call("Plugin.HjConnRWWrite", b, &n); err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+func (w *hijackedConn) Read(b []byte) (int, error) {
+	var data []byte
+	if err := w.client.Call("Plugin.HjConnRead", len(b), &data); err != nil {
+		return 0, nil
```

Comment:
Intentional to return `nil` for the `error` here?

Reply:
Definitely not. Good catch! :+1: